src/atomic_shared_ptr.h

```diff
@@ -292,64 +317,130 @@ void AtomicSharedPtr<T>::store(T *data) {
-                }
-                continue;
+        size_t expectedPackedPtr = holder.knownValue;
+
+        int addCredits = 0;
+        while (true)
+        {
+            int localRefCount = expectedPackedPtr & MAGIC_MASK;
+
+            int diff = localRefCount - addCredits;
+            if (diff > 1)
+            {
+                addCredits += diff;
+                auto before = controlBlock->refCount.fetch_add(diff);
+				FAST_LOG(Operation::CompareAndSwap_Inc, (expectedPackedPtr & ~MAGIC_MASK) | (before + diff));
+                assert(before > 0);
             }
+
             assert((expectedPackedPtr >> MAGIC_LEN) != (desiredPackedPtr >> MAGIC_LEN));
-            if (packedPtr.compare_exchange_weak(expectedPackedPtr, desiredPackedPtr)) {
-                FAST_LOG(Operation::GetInCAS, expectedPackedPtr);
-                newOne.controlBlock = nullptr;
-                assert((expectedPackedPtr >> MAGIC_LEN) == holdedPtr);
-                destroyOldControlBlock(expectedPackedPtr);
-                return true;
+            if (packedPtr.compare_exchange_weak(expectedPackedPtr, desiredPackedPtr) == false)
+            {
+                if (holdedPtr != (expectedPackedPtr >> MAGIC_LEN))
+                	break;
+
+                continue;
             }
+
+            FAST_LOG(Operation::GetInCAS, expectedPackedPtr);
+            newOne.controlBlock = nullptr;
+            assert((expectedPackedPtr >> MAGIC_LEN) == holdedPtr);
+
+        	FAST_LOG(Operation::CompareAndSwap_Diff, reinterpret_cast<size_t>(holder.getControlBlock()) << MAGIC_LEN | diff);
+
+            int correctionForGlobalCounter = addCredits - localRefCount;
+            assert(correctionForGlobalCounter >= -1 && "The global ref count must be either on point, overshoot some, or exactly one down for `holder` it did not count in");
+
+            int finalCorrection = correctionForGlobalCounter;
+            finalCorrection += /*Yielding `holder` ownership*/ 1;
+            finalCorrection += /*Yielding `this` ownership*/ 1;
+        	FAST_LOG(Operation::CompareAndSwap_Diff2, reinterpret_cast<size_t>(holder.getControlBlock()) << MAGIC_LEN | finalCorrection);
+			destroyOldControlBlock(expectedPackedPtr, finalCorrection);
+            holder.foreignPackedPtr = nullptr;
+
+            return true;
+        }
+
+        if (addCredits)
+        {
+	        assert(addCredits > 0);
+
+	        auto before = controlBlock->refCount.fetch_sub(addCredits);
+	        assert
+			(
+	            before > 0 && 
+	            "At least `holder` must still have 1 active count by now"
+	            "That count must be now promoted to global by someone who exchanged block in `this`, otherwise we'd be still CASing"
+	            "Note that even if `this` still holds that same block, it would have to hold one global count of its own, thus we still can't observe 0 here"
+	        );
         }
     }
 
+    if constexpr (IsStoreOp)
+    {
+        goto Reacquire;
```

Comment:
I really don't like goto. Maybe leave it as tail recursion optimization? 

```__attribute__((musttail)) return compareExchangeImpl(expected, std::move(newOne));```

Or even just a while(true) cycle is fine as well.